File: 1654-minimum-jumps-to-reach-home/1654-minimum-jumps-to-reach-home.py
class Solution:
    def minimumJumps(self, forbidden: List[int], a: int, b: int, x: int) -> int:
        s = set(forbidden)
        # sys.setrecursionlimit(10**5)
        # @lru_cache(None)
        dp = [[-1,-1] for i in range(6001)]
        def compute(position,back):
            
            if position==x:
                return 0
            
            if position<0:
                return float('inf')
            if position>6000:
                return float('inf')
            if back>=2:
                return float('inf')
            if position in s:
                return float('inf')
            
            if dp[position][back]!=-1:
                return dp[position][back]
            
            right = 1 + compute(position + a,0)
            dp[position][back] = right
            if not back:
                right =  min(right,1 + compute(position - b,back + 1))
            dp[position][back] = right
            return right
        # left = compute(0,0)
        # right= compute(0,0)
        # print(left,right)
        # res  = min(left,right)
        res = compute(0,0)
        return -1 if res == float('inf') else res

File: 1654-minimum-jumps-to-reach-home/1654-minimum-jumps-to-reach-home.cpp
class Solution {
public:
unordered_map<int,int>mp;
int dp[6001][2];
int fun(int i, int c,int a, int b, int x)
{
if(i==x)return 0;
if(i<0 or i>6000 or c>=2 or mp.find(i)!=mp.end())return 1e9;

    if(dp[i][c]!=-1)return dp[i][c];
    
    dp[i][c]=1+fun(i+a,0,a,b,x);
    if(!c)
        dp[i][c]=min(dp[i][c],1+fun(i-b,c+1,a,b,x));
    
    return dp[i][c];
}
int minimumJumps(vector<int>& forbidden, int a, int b, int x) {
    for(int i=0;i<forbidden.size();i++)
        mp[forbidden[i]]=1;
    memset(dp,-1,sizeof(dp));
    
    int res=fun(0,0,a,b,x);
    if(res>=1e9)return -1;
    return res;
}
};